SCHEMA.md

```diff
@@ -0,0 +1,211 @@
+
+And an extended primitive type example: `ip_port`
+
+```
+-type ip_port() :: tuple().
+-typerefl_from_string({ip_port/0, this_module, to_ip_port}).
+to_ip_port(String) ->
+    case string:tokens(String) of
+        ....
+    end.
+```
+
+### Complex types
+
+HOCON schema supports 3 different complext types: `struct`, `array`, and `union`.
+NOTE: to make it easier for future extensions, it's recommended to use `hoconsc` module APIs to define schema.
+
+#### Structs
+
+NOTE: HOCON schema does not support non-struct root level data types. e.g. it is not allowed to
+define a root level schema with just a `integer()` type.
+
+Structs consist of data fields, which can be defined using `hocon_schema` behaviour callbacks.
+
+* `structs/0`: This callback returns all the root level namespaces.
+* `fields/1`: THis callback returns the schema for each data field (in a list, so order matters).
+
+For example, to define a struct named `foo` having one integer field, the schema module may look like:
+
+```
+-export([structs/0, fields/1]).
+structs() -> ["foo"]. %% 'exported' root names
+fields("foo") -> [{"field1", typerefl:integer()}].
+```
+
+In this case, the schema for use in `hocon_schema` APIs is the module name. There is another way to
+define a struct as a Erlang `map()`, so we do not have to implement the behaviour callbacks
+(this is however mostly for test cases):
+
+```
+#{structs => ["foo"], %% 'exported' root names
+  fields => #{"foo" => [{"field1", typerefl:integer()}]}
+ }
+```
+
+##### Struct references
+
+In order to promote code abstraction and prevent copy-paste as much as possible,
+in HOCON schema, there is no way to define structs nested (child struct nested in a parent struct).
+The parent-children relationship has to be defined as struct 'referencing'.
+
+e.g. if the type of parent-struct's field is another struct, the field's type should be defined as:
+
+```
+[ ...,
+  {field_N, hoconsc:ref("field_struct_name")},
+  ...
+].
+```
+
+##### Virtual struct root
+
+The root struct name exported in the `structs/0` API serves as top level struct's field names.
+like `listener`, `zone` and `broker` in `etc/emqx.conf`.
+
+#### Arrays
+
+Array is a sequence of other types which is defined as `{array, Type}`.
+
+#### Unions
+
+A union type is in some contexts `one_of` types.
+When data is validated against the schema (recursively), the code enumerates
+the union member types in the defined order until the given data matches any of the union member.
+
+### Config generation
+
+When starting a Erlang node it requires a system configuration file, (usually named `sys.config`),
+see [Erlang doc](https://erlang.org/doc/man/config.html#sys.config) for more details.
+
+When using HOCON config format, we need a tool to tranfrom a HOCON file to a config file of `sys.config` format.
+`hocon_schema` is such a tool.
+
+### Config mapping
+
+As introduced above, Erlang requires a `sys.config` to bootstrap, the content of this config is essentially
```

Comment:
I think `sys.config` is optional for erlang.  You may say some application requires 'sys.config' to bootstrap.

Reply:
I have rephrased it in the above section with a 'usually' word, and removed 'requires ...' here.